// local-error-formatter/sparkpost-formatter.js
/**
 * Format the error according to SparkPost
 *
 * @function
 * @param {Object} err - Error from SparkPost
 * @returns {Object} - Formatted error object to be passed to formatError
 */
function sparkPostError(err) {
  // Defaults are set to fit the majority of use cases
  const error = {
    level: 'warn',
    info: err.errors,
    message: err.message,
    statusCode: err.statusCode,
    jsonResponse: {
      status: 'fail',
      name: err.name,
      message: err.message,
      data: err.errors.reduce((result, obj) => {
        const map = result;
        map.info = obj.message;
        return map;
      }, {}),
    },
  };

  switch (err.statusCode.toString()) {
    case '400':
    case '401':
    case '403':
    case '404':
    case '405':
    case '409':
    case '415':
    case '422':
    case '429':
      error.level = 'error';
      error.jsonResponse.status = 'error';
      error.jsonResponse.data = null;

      return error;
    case '500':
    case '503':
      error.level = 'error';
      error.jsonResponse.status = 'error';
      error.jsonResponse.data = null;

      return error;
    default:
      return error;
  }
}

module.exports = sparkPostError;
