// local-error-formatter/stripe-formatter.js
/**
 * Format the error according to Stripe Payments API
 *
 * @function
 * @param {Object} err - Error from Stripe
 * @returns {Object} - Formatted error object to be passed to formatError
 */
function stripeError(err) {
  // Defaults are set to fit the majority of use cases
  const error = {
    level: 'warn',
    info: {
      chargeId: err.charge,
      code: err.code,
      decline_code: err.decline_code,
      param: err.param,
    },
    message: err.message,
    statusCode: 400,
    jsonResponse: {
      status: 'fail',
      name: err.type,
      data: { payment: err.message },
    },
  };

  switch (err.type) {
    case 'StripeCardError':
    case 'RateLimitError':
    case 'StripeInvalidRequestError':
    case 'StripeAPIError':
    case 'StripeConnectionError':
    case 'StripeAuthenticationError':
      return error;
    default:
      error.statusCode = 500;
      error.level = 'error';
      error.jsonResponse.status = 'error';
      error.jsonResponse.message = err.message;

      return error;
  }
}

module.exports = stripeError;
