// local-models/user/user-validations.js
const { validators } = require('validate.js');

/**
 * Password must contain at least 1 number
 */
validators.oneNumber = (value, options, key, attributes) => {
  if (attributes.password && !(/^(?=.*[0-9])(.+)$/).test(attributes.password)) {
    return 'must contain at least 1 number';
  }

  return null;
};

/**
 * Password must contain at least 1 lowercase letter
 */
validators.oneLowercaseLetter = (value, options, key, attributes) => {
  if (attributes.password && !(/^(?=.*[a-z])(.+)$/).test(attributes.password)) {
    return 'must contain at least 1 lowercase letter';
  }

  return null;
};

/**
 * Password must contain at least 1 uppercase letter
 */
validators.oneUppercaseLetter = (value, options, key, attributes) => {
  if (attributes.password && !(/^(?=.*[A-Z])(.+)$/).test(attributes.password)) {
    return 'must contain at least 1 uppercase letter';
  }

  return null;
};

module.exports = {
  first_name: {
    format: {
      pattern: /[A-Za-z'-.]+/,
      message: 'must contain only letters and \'-.',
    },
  },
  last_name: {
    format: {
      pattern: /[A-Za-z'-.]+/,
      message: 'must contain only letters and \'-.',
    },
  },
  email: { email: true },
  password: {
    oneNumber: true,
    oneUppercaseLetter: true,
    oneLowercaseLetter: true,
    length: {
      minimum: 10,
      maximum: 128,
      tooShort: 'must be at least 10 characters',
      tooLong: 'must be less than or equal to 128 characters',
    },
  },
};
