// local-error-handler/formatters/stripe-formatter.js
const objectPath = require('object-path');

/**
 * Format the error according to Stripe Payments API
 *
 * @function
 * @param {Object} err - Error from Stripe
 * @returns {Object} - Formatted error object to be passed to formatError
 */
function stripeError(err) {
  // Set a default properties
  const statusCode = objectPath.get(err, 'statusCode', '500');
  const code = objectPath.get(err, 'code', 'STRIPE_ERROR');
  const param = objectPath.get(err, 'param', undefined);

  // Defaults are set to fit the majority of use cases
  const error = {
    level: 'warn',
    message: err.message,
    statusCode,
    jsonResponse: [{
      statusCode: statusCode.toString(),
      message: err.message,
      code: code.toUpperCase(),
      meta: { param },
    }],
  };

  switch (err.type) {
    case 'StripeCardError':
    case 'StripeInvalidRequestError':
    case 'StripeAuthenticationError':
      return error;
    case 'RateLimitError':
    case 'StripeConnectionError':
    case 'StripeAPIError':
    default:
      error.level = 'error';

      return error;
  }
}

module.exports = stripeError;
