// local-logger/local-logger.js
const bunyan = require('bunyan');
const PrettyStream = require('bunyan-prettystream');
const config = require('config');
const path = require('path');

const prettyStdOut = new PrettyStream();
const ringbuffer = new bunyan.RingBuffer({ limit: 100 });

const logger = bunyan.createLogger({
  name: 'node-api',
  level: (config.logger && config.logger.level) || 'info',
  serializers: {
    err: bunyan.stdSerializers.err,
    req(req) {
      if (!req) {
        return (false);
      }

      const whitelistedHeaders = () => {
        const headers = Object.assign({}, req.headers);
        delete headers.authorization;
        return headers;
      };

      return ({
        query: req.query,
        params: req.params,
        method: req.method,
        url: req.url,
        headers: whitelistedHeaders(),
        httpVersion: req.httpVersion,
      });
    },
    res(res) {
      if (!res) {
        return (false);
      }

      return ({
        statusCode: res.statusCode,
        headers: res._headers,
      });
    },
  },
  streams: [
    {
      stream: prettyStdOut.pipe(process.stdout),
      level: 'debug',
    },
    {
      type: 'rotating-file',
      path: path.join(process.cwd(), '/logs/app.log'),
      period: '1d', // daily rotation
      count: 4, // keep 4 back copies
    },
    {
      level: 'trace',
      type: 'raw', // use 'raw' to get raw log record objects
      stream: ringbuffer,
    },
  ],
});

// Expose the ringbugger for when we need it
logger.ringbuffer = ringbuffer;

module.exports = logger;
